// @huddle01/huddle01-client/dist/hooks/index.mjs
import {
  usePeerCamTrack_default
} from "../chunk-3RCU5NAL.mjs";
import {
  usePeersData_default
} from "../chunk-YKNBSP2C.mjs";
import {
  useMeCamTrack_default
} from "../chunk-GLBOZTYT.mjs";
import {
  useMediaDevices_default
} from "../chunk-C6J2EIDU.mjs";
import {
  HuddleClientContext
} from "../chunk-OEW27RTB.mjs";
import {
  store_default
} from "../chunk-DM7RIPEQ.mjs";
import "../chunk-N5JBSBUH.mjs";
import "../chunk-VAZJKIJW.mjs";
import "../chunk-OJOQ4W4X.mjs";
import "../chunk-DNIVQHGU.mjs";
import "../chunk-CRMTBXWZ.mjs";
import "../chunk-UL7IYC3S.mjs";
import "../chunk-LNN4OEFN.mjs";
import "../chunk-SO4JIC34.mjs";
import "../chunk-D6AGFJZ3.mjs";
import "../chunk-APOZGH7P.mjs";
import "../chunk-E3J4KYLN.mjs";
import "../chunk-RWDW3Z55.mjs";
import "../chunk-4GVG6VSP.mjs";
import "../chunk-45YZ7PTW.mjs";
import "../chunk-LGGEZBM7.mjs";
import "../chunk-ZTBTXBLL.mjs";
import "../chunk-EIL454WZ.mjs";
import "../chunk-PGV2CJQO.mjs";
import "../chunk-LRNXPFD3.mjs";
import "../chunk-FEVUFAUA.mjs";
import "../chunk-GAPPCOE7.mjs";
import "../chunk-7GGWN2HC.mjs";
import "../chunk-7JMHGAJI.mjs";
import "../chunk-SQKZRIEU.mjs";
import "../chunk-IBGL6YYT.mjs";

// src/hooks/index.ts
import { useContext } from "react";
var useHuddleClientContext = () => useContext(HuddleClientContext);
export {
  useHuddleClientContext,
  store_default as useHuddleStore,
  useMeCamTrack_default as useMeCamTrack,
  useMediaDevices_default as useMediaDevices,
  usePeersData_default as usePeersData,
  usePeerCamTrack_default as usePeersTracks
};
