// @huddle01/huddle01-client/dist/chunk-DM7RIPEQ.mjs
import {
  createTestSlice_default
} from "./chunk-N5JBSBUH.mjs";
import {
  createMediaSlice_default
} from "./chunk-VAZJKIJW.mjs";
import {
  createMicSlice_default
} from "./chunk-OJOQ4W4X.mjs";
import {
  createNetworkSlice_default
} from "./chunk-DNIVQHGU.mjs";
import {
  createPeerNameSlice_default
} from "./chunk-CRMTBXWZ.mjs";
import {
  createPeersSlice_default
} from "./chunk-UL7IYC3S.mjs";
import {
  createRecordingSlice_default
} from "./chunk-LNN4OEFN.mjs";
import {
  createRoomStateSlice_default
} from "./chunk-SO4JIC34.mjs";
import {
  createSpeakerSlice_default
} from "./chunk-D6AGFJZ3.mjs";
import {
  createFigmentSlice_default
} from "./chunk-APOZGH7P.mjs";
import {
  createGridSlice_default
} from "./chunk-RWDW3Z55.mjs";
import {
  createHostControlsSlice_default
} from "./chunk-4GVG6VSP.mjs";
import {
  createLastNSlice_default
} from "./chunk-45YZ7PTW.mjs";
import {
  createLayoutSlice_default
} from "./chunk-LGGEZBM7.mjs";
import {
  createLiveStreamingSlice_default
} from "./chunk-ZTBTXBLL.mjs";
import {
  createLobbyPeersSlice_default
} from "./chunk-EIL454WZ.mjs";
import {
  createMeSlice_default
} from "./chunk-PGV2CJQO.mjs";
import {
  createActiveSpeakerSlice_default
} from "./chunk-FEVUFAUA.mjs";
import {
  createChatSlice_default
} from "./chunk-7GGWN2HC.mjs";

// src/store/store.ts
import create from "zustand";
import createVanilla from "zustand/vanilla";
import { devtools } from "zustand/middleware";
var createRootSlice = (set, get) => ({
  ...createGridSlice_default(set, get),
  ...createMeSlice_default(set, get),
  ...createPeersSlice_default(set, get),
  ...createTestSlice_default(set, get),
  ...createLastNSlice_default(set, get),
  ...createMediaSlice_default(set, get),
  ...createFigmentSlice_default(set, get),
  ...createRecordingSlice_default(set, get),
  ...createChatSlice_default(set, get),
  ...createLobbyPeersSlice_default(set, get),
  ...createRoomStateSlice_default(set, get),
  ...createNetworkSlice_default(set, get),
  ...createHostControlsSlice_default(set, get),
  ...createLiveStreamingSlice_default(set, get),
  ...createSpeakerSlice_default(set, get),
  ...createMicSlice_default(set, get),
  ...createActiveSpeakerSlice_default(set, get),
  ...createLayoutSlice_default(set, get),
  ...createPeerNameSlice_default(set, get)
});
var store = createVanilla(
  devtools(createRootSlice, {
    name: "huddle01-client",
    anonymousActionType: "action",
    enabled: process.env.NODE_ENV === "development"
  })
);
var useRootStore = create(store);
var useHuddleStore = useRootStore;
var { getState, setState } = useRootStore;
var store_default = store;

export {
  useRootStore,
  useHuddleStore,
  getState,
  setState,
  store_default
};
