// @huddle01/huddle01-client/dist/chunk-SQKZRIEU.mjs
// src/constants.ts
var VIDEO_CONSTRAINS = {
  qvga: { width: { ideal: 320 }, height: { ideal: 240 } },
  vga: { width: { ideal: 640 }, height: { ideal: 480 } },
  hd: { width: { ideal: 1280 }, height: { ideal: 720 } }
};
var PC_PROPRIETARY_CONSTRAINTS = {
  optional: [{ googDscp: true }]
};
var WEBCAM_SIMULCAST_ENCODINGS = [
  {
    scaleResolutionDownBy: 4,
    maxBitrate: 3e5,
    rid: "r0",
    scalabilityMode: "S1T3"
  },
  {
    scaleResolutionDownBy: 2,
    maxBitrate: 6e5,
    rid: "r1",
    scalabilityMode: "S1T3"
  },
  {
    scaleResolutionDownBy: 1,
    maxBitrate: 9e6,
    rid: "r2",
    scalabilityMode: "S1T3"
  }
];
var WEBCAM_KSVC_ENCODINGS = [{ scalabilityMode: "S3T3_KEY" }];
var SCREEN_SHARING_SIMULCAST_ENCODINGS = [
  { dtx: true, maxBitrate: 15e5 },
  { dtx: true, maxBitrate: 6e6 }
];
var SCREEN_SHARING_SVC_ENCODINGS = [
  { scalabilityMode: "S3T3", dtx: true }
];

export {
  VIDEO_CONSTRAINS,
  PC_PROPRIETARY_CONSTRAINTS,
  WEBCAM_SIMULCAST_ENCODINGS,
  WEBCAM_KSVC_ENCODINGS,
  SCREEN_SHARING_SIMULCAST_ENCODINGS,
  SCREEN_SHARING_SVC_ENCODINGS
};
